Change `qzdiv` to `ordschur` in gensys.jl

pearlzli opened the pull request:
Closes #1.

Note for future DSGE package developers: the version of `gensys` in DSGE.jl as of today converts the input matrices Γ0 and Γ1 to complex before calling `schurfact!`. This was done to parallel the default behavior in Matlab `qz`, which is to carry out the complex Schur decomposition. However, I don't do this conversion to complex in this pull request in the pursuit of a more Julian approach to type-stability. Thus when we eventually move to calling the QuantEcon version of `gensys`, we'll need to explicitly call `gensys(complex(Γ0), complex(Γ1), C, ...)` when solving the model. Otherwise we get a SingularException at the line `G0I = inv(G0)`.

pearlzli committed:
Account for gensys side effects

- Rename `gensys` to `gensys!`
- Add wrapper methods `gensys` that pass in copies of matrices to `gensys!`

pearlzli committed:
Change qzdiv to ordschur

pearlzli committed:
Remove qzdiv and qzswitch

pearlzli committed:
Change to lowercase c in gensys wrapper methods

pearlzli committed:
Add ordered_qz.jl for v0.3 compatibility

pearlzli committed:
Add ordered_qz.jl tests

sglyon commented on src/gensys.jl:
```diff
@@ -88,23 +89,28 @@ end
 
+# Wrapper methods ensuring no side effects on input arguments
+gensys(Γ0, Γ1, c, ψ, π) = gensys!(copy(Γ0), copy(Γ1), copy(c), copy(ψ), copy(π))
+gensys(Γ0, Γ1, c, ψ, π, div) = gensys!(copy(Γ0), copy(Γ1), copy(c), copy(ψ), copy(π), div)
```
Do all these need to be copied? Meaning do all of them get mutated within `gensys!`?

pearlzli replied:
That's a very good point - it looks like only Γ0 and Γ1 are changed within `gensys`, and there only by the call to `schurfact!`. I'll change that to the non-mutating `schurfact` (which itself will copy Γ0 and Γ1).

spencerlyon2 commented:
This is awesome. Wonderful to have the back ported `ordschur` methods here and have some tests!

Will merge after we figure out the answer to the question in the inline comment above

pearlzli committed:
Reduce unnecessary copying in gensys

- Change `schurfact!` to `schurfact`
- Remove `gensys` wrapper methods passing in copies of all input matrices
- Rename `gensys!` back to `gensys`

spencerlyon2 commented:
Excellent, thank you!
